src/hugr/replace.rs

```diff
@@ -0,0 +1,14 @@
+//! Rewrite operations on the HUGR.
+
+#[allow(clippy::module_inception)] // TODO: Rename?
+pub mod rewrite;
+use crate::Hugr;
+pub use rewrite::{OpenHugr, Rewrite, RewriteError};
+
+/// An operation that can be applied to mutate a Hugr
+pub trait RewriteOp<E> {
```

Comment:
We should make this a [sealed trait](https://rust-lang.github.io/api-guidelines/future-proofing.html#sealed-traits-protect-against-downstream-implementations-c-sealed), since we don't want user writing their own rewrite ops (for now at least).

Reply:
I am not entirely convinced here. I think I could be persuaded, but especially given making a sealed trait involves playing around with some mildly complex arrangement of namespaces etc. (not just added a `sealed` between `pub` and `trait`, say), I'm still pondering what this achieves and why.

* Implementing the trait doesn't give you any power to see inside the Hugr. That power only comes from being defined inside the hugr module.
* The only point of implementing the trait is to be able to do `h.apply_rewrite(r)` rather than `r.apply(&mut h)`. Is that such a signal that we want to restrict it to "our" classes? It may be....

In practice many transformations one might want to apply to a Hugr are going to involve a series of the Rewrite's that we will define here. For example, an outlining, following by a replace, followed by an inlining. If we seal the trait, such transformations are distinguished from the "primitives" that we provide. Is that useful? *Maybe*. But it is certainly less compositional (one loses the ability to parameterize code over Rewrites including custom/compound ones, for example)....